crates/wasmparser/src/validator/component.rs

```diff
@@ -1041,13 +1074,573 @@ impl ComponentState {
         offset: usize,
     ) -> Result<()> {
         let rep = self.check_local_resource(resource, types, offset)?;
-        let func_ty = FuncType::new([ValType::I32], [rep]);
-        let core_ty = SubType::func(func_ty, false);
-        let id = types.intern_sub_type(core_ty, offset);
+        let id = types.intern_func_type(FuncType::new([ValType::I32], [rep]), offset);
+        self.core_funcs.push(id);
+        Ok(())
+    }
+
+    pub fn task_backpressure(
+        &mut self,
+        types: &mut TypeAlloc,
+        offset: usize,
+        features: &WasmFeatures,
+    ) -> Result<()> {
+        if !features.component_model_async() {
+            bail!(
+                offset,
+                "`task.backpressure` requires the component model async feature"
+            )
+        }
+
+        self.core_funcs
+            .push(types.intern_func_type(FuncType::new([ValType::I32], []), offset));
+        Ok(())
+    }
+
+    pub fn task_return(
+        &mut self,
+        type_index: u32,
+        types: &mut TypeAlloc,
+        offset: usize,
+        features: &WasmFeatures,
+    ) -> Result<()> {
+        if !features.component_model_async() {
+            bail!(
+                offset,
+                "`task.return` requires the component model async feature"
+            )
+        }
+
+        let id = self.type_id_at(type_index, offset)?;
+        let Some(SubType {
+            composite_type:
+                CompositeType {
+                    inner: CompositeInnerType::Func(_),
```

Comment:
Yeah, I've struggled with this myself, and indeed my first draft (written before a spec existed) used a component function type instead of a core function type.  After all, you can always derive the latter from the former.  Luke and I discussed it and landed on the current scheme, but I don't remember the details as to why.  Anyway, yes, please open an issue upstream about it.

Reply:
I chatted with Luke about this, and he'll update the spec to use the component-level return type instead of the core function type for `task.return` declarations, then I'll update the implementation to match.